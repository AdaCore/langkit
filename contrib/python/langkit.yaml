lkt_spec:
  entry_point: nodes.lkt
  source_dirs: [.]
library:
  language_name: Python
warnings:
  undocumented-nodes: false
