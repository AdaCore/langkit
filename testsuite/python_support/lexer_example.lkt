lexer foo_lexer {

    @trivia()
    whitespace <- p"[ \n\r\t]+"

    family alphanumericals {
        def <- "def"
        var <- "var"
    }

    family alphanumericals {
        error <- "error"
        example <- "example"
        null_tok <- "null"
    }


    family punctuation {
        comma <- ","
        dot <- "."
        semicolon <- ";"
        l_par <- "("
        r_par <- ")"
        l_brace <- "{"
        r_brace <- "}"
        equal <- "="
        plus <- "+"
    }

    minus <- "-"
    less_than <- "<"
    less_than_or_equal <- "<="
    greater_than <- ">"
    greater_than_or_equal <- ">="

    @unparse_spacing(with=alphanumericals)
    family alphanumericals {
        number <- p"[0-9]+|0x[0-9a-f]+"
        @symbol()
        identifier <- p"[a-zA-Z_][a-zA-Z0-9_]*"
    }

    string <- p"\"[^\"]*\""

    family comments {
        @unparse_newline_after
        @trivia()
        comment <- p"#(.?)+"
    }

}
