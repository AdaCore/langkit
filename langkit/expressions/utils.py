from __future__ import annotations

import langkit.expressions as E


def inc_ref(var: E.VariableExpr | E.LocalVars.LocalVar) -> str:
    """
    Shortcut for expression rendering.

    If ``var`` is a ref-counted variable, return Ada code to inc-ref it
    (one statement including trailing semi-colon). Otherwise, return an empty
    string.
    """
    var_name = (
        var.codegen_name if isinstance(var, E.LocalVars.LocalVar) else var.name
    )
    return f"Inc_Ref ({var_name});" if var.type.is_refcounted else ""


def assign_var(
    var: E.VariableExpr | E.LocalVars.LocalVar,
    expr: str,
    requires_incref: bool = True,
) -> str:
    """
    Shortcut for expression rendering.

    Return Ada code to assign the given `expr` to the given `var`.  If
    `requires_incref` is True and `var` is a ref-counted variable, also emit
    code to inc-ref it. The result is a sequence of Ada statements, including
    trailing semi-colon.
    """
    var_name = (
        var.codegen_name if isinstance(var, E.LocalVars.LocalVar) else var.name
    )
    return f"{var_name} := {expr}; {inc_ref(var) if requires_incref else ''}"


def array_aggr(exprs: list[str]) -> str:
    """
    Shortcut to format an array aggregate expression.
    """
    return (
        "({})".format(", ".join(["{}"] * len(exprs)))
        if exprs
        else "(1 .. 0 => <>)"
    )
