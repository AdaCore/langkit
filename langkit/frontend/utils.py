from __future__ import annotations

from collections import OrderedDict
import os.path

from langkit.compile_context import CompileCtx
from langkit.config import LktSpecConfig
from langkit.diagnostics import (
    Location,
    error,
    errors_checkpoint,
    non_blocking_error,
)
import langkit.names as names

import liblktlang as L


def name_from_lower(ctx: CompileCtx, kind: str, id: L.Id) -> names.Name:
    """
    Validate "id" as a lower-case name and return the corresponding ``Name``
    instance.
    """
    try:
        names.check_common(id.text)
    except ValueError as exc:
        error(str(exc), location=id)
    try:
        names.check_lower(id.text)
    except ValueError:
        error(f"lower case expected for {kind} names", location=id)
    return names.Name.from_lower(id.text)


def name_from_camel(ctx: CompileCtx, kind: str, id: L.Id) -> names.Name:
    """
    Validate "id" as a camel-case name and return the corresponding ``Name``
    instance.
    """
    try:
        names.check_common(id.text)
    except ValueError as exc:
        error(str(exc), location=id)
    try:
        names.check_camel(id.text)
    except ValueError as exc:
        error(f"camel case expected for {kind} names ({exc})", location=id)
    return names.Name.from_camel(id.text)


def arg_name_from_expr(e: L.Expr) -> L.Id:
    """
    Return the argument name for a given argument value expression.

    This assumes that ``e`` is the ``f_value`` argument of a named ``Argument``
    node, and return the ``f_name`` field for that argument.
    """
    arg = e.parent
    assert isinstance(arg, L.Argument)
    assert arg.f_name is not None
    return arg.f_name


def lkt_doc(decl: L.Decl) -> str:
    """
    Return the documentation attached to the ``decl`` node. This is an empty
    string if the docstring is missing.

    :param decl: Declaration to process.
    """
    from langkit.frontend.static import denoted_str

    full_decl = decl.parent
    assert isinstance(full_decl, L.FullDecl)
    return "" if full_decl.f_doc is None else denoted_str(full_decl.f_doc)


def load_lkt(config: LktSpecConfig) -> list[L.AnalysisUnit]:
    """
    Load a Lktlang source file and return the closure of Lkt units referenced.
    Raise a DiagnosticError if there are parsing errors.

    :param lkt_file: Name of the file to parse.
    """
    units_map = OrderedDict()
    diagnostics = []

    def process_unit(unit: L.AnalysisUnit) -> None:
        if unit.filename in units_map:
            return

        # Register this unit and its diagnostics
        units_map[unit.filename] = unit
        for d in unit.diagnostics:
            diagnostics.append((unit, d))

        # Recursively process the units it imports. In case of parsing error,
        # just stop the recursion: the collection of diagnostics is enough.
        if not unit.diagnostics:
            assert isinstance(unit.root, L.LangkitRoot)
            import_stmts = list(unit.root.f_imports)
            for imp in import_stmts:
                process_unit(imp.p_referenced_unit)

    # Give Liblktlang access to the Lkt files to analyze
    old_path = os.environ.get("LKT_PATH", "")
    os.environ["LKT_PATH"] = os.path.pathsep.join(config.source_dirs)

    # Load ``lkt_file`` and all the units it references, transitively
    ctx = L.AnalysisContext(unit_provider=L.UnitProvider.from_lkt_path())
    process_unit(ctx.get_from_file(config.entry_point))

    os.environ["LKT_PATH"] = old_path

    # Forward potential lexing/parsing errors to our diagnostics system
    for u, d in diagnostics:
        non_blocking_error(
            d.message, Location.from_sloc_range(u, d.sloc_range)
        )
    errors_checkpoint()
    return list(units_map.values())
