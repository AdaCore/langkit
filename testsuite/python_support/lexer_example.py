from __future__ import absolute_import, division, print_function

from langkit.lexer import (Lexer, LexerToken, Literal, Pattern, TokenFamily,
                           WithSymbol, WithText, WithTrivia)


class Token(LexerToken):
    Def = WithText()
    Var = WithText()
    Error = WithText()
    Example = WithText()
    Null = WithText()

    Comma = WithText()
    Dot = WithText()
    Semicolon = WithText()
    LPar = WithText()
    RPar = WithText()
    LBrace = WithText()
    RBrace = WithText()
    Equal = WithText()
    Plus = WithText()
    Minus = WithText()
    LessThan = WithText()

    Number = WithText()
    Identifier = WithSymbol()
    String = WithText()

    Comment = WithTrivia()
    Whitespace = WithTrivia()

    Alphanumericals = TokenFamily(Def, Error, Example, Null, Number,
                                  Identifier)
    Punctuation = TokenFamily(Comma, Dot, Semicolon, LPar, RPar, LBrace,
                              RBrace, Equal, Plus)
    Comments = TokenFamily(Comment)


foo_lexer = Lexer(Token)
foo_lexer.add_rules(
    (Pattern(r'[ \n\r\t]+'), Token.Whitespace),

    (Literal('def'),         Token.Def),
    (Literal('var'),         Token.Var),
    (Literal('error'),       Token.Error),
    (Literal('example'),     Token.Example),
    (Literal('null'),        Token.Null),

    (Literal(','),           Token.Comma),
    (Literal('.'),           Token.Dot),
    (Literal(';'),           Token.Semicolon),
    (Literal('('),           Token.LPar),
    (Literal(')'),           Token.RPar),
    (Literal('{'),           Token.LBrace),
    (Literal('}'),           Token.RBrace),
    (Literal('='),           Token.Equal),
    (Literal('+'),           Token.Plus),
    (Literal('-'),           Token.Minus),
    (Literal('<'),           Token.LessThan),

    (Pattern('[0-9]+'),      Token.Number),
    (Pattern('[a-zA-Z_][a-zA-Z0-9_]*'), Token.Identifier),
    (Pattern(r'\"[^\"]*\"'), Token.String),

    (Pattern('#(.?)+'), Token.Comment),
)
foo_lexer.add_spacing((Token.Alphanumericals, Token.Alphanumericals))
foo_lexer.add_newline_after(Token.Comment)
