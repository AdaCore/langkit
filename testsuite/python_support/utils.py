from __future__ import absolute_import, division, print_function

import os
import os.path
import shutil
import subprocess
import sys

import langkit
import langkit.compile_context
from langkit.compile_context import CompileCtx
from langkit.diagnostics import DiagnosticError, WarningSet
from langkit.libmanage import ManageScript

from testsuite_support.valgrind import valgrind_cmd


default_warning_set = WarningSet()

# We don't want to be forced to provide dummy docs for nodes and public
# properties in testcases.
default_warning_set.disable(WarningSet.undocumented_nodes)
default_warning_set.disable(WarningSet.undocumented_public_properties)

pretty_print = bool(int(os.environ.get('LANGKIT_PRETTY_PRINT', '0')))

project_template = """
with "libfoolang";

project Gen is
    for Languages use ("Ada");
    for Source_Dirs use (".");
    for Object_Dir use "obj";
    for Main use ({main_sources});

    package Compiler is
        for Default_Switches ("Ada") use
          ("-g", "-O0", "-gnata", "-gnatwae", "-gnatyg");
    end Compiler;
end Gen;
"""


# Determine where to find the root directory for Langkit sources
langkit_root = os.environ.get('LANGKIT_ROOT_DIR')
if not langkit_root:
    test_dir = os.path.dirname(os.path.abspath(__file__))
    testsuite_dir = os.path.dirname(test_dir)
    langkit_root = os.path.dirname(testsuite_dir)


def prepare_context(grammar, lexer=None, warning_set=default_warning_set,
                    symbol_canonicalizer=None):
    """
    Create a compile context and prepare the build directory for code
    generation.

    :param langkit.parsers.Grammar grammar: The language grammar to use for
        this context.

    :param langkit.lexer.Lexer lexer: The language lexer to use for this
        context.

    :param WarningSet warning_set: Set of warnings to emit.

    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    # Have a clean build directory
    if os.path.exists('build'):
        shutil.rmtree('build')
    os.mkdir('build')

    # Try to emit code
    ctx = CompileCtx(lang_name='Foo', lexer=lexer, grammar=grammar,
                     symbol_canonicalizer=symbol_canonicalizer)
    ctx.warnings = warning_set
    ctx.pretty_print = pretty_print

    return ctx


def emit_and_print_errors(grammar, lexer=None,
                          warning_set=default_warning_set,
                          generate_unparser=False, symbol_canonicalizer=None):
    """
    Compile and emit code for CTX. Return whether this was successful.

    :param langkit.parsers.Grammar grammar_fn: The language grammar to use.

    :param langkit.lexer.Lexer lexer: The lexer to use along with the grammar.
        Use `lexer_example.foo_lexer` if left to None.

    :param WarningSet warning_set: Set of warnings to emit.

    :param bool generate_unparser: Whether to generate unparser.

    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.

    :rtype: bool
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    try:
        ctx = prepare_context(grammar, lexer, warning_set,
                              symbol_canonicalizer=symbol_canonicalizer)
        ctx.emit('build', generate_unparser=generate_unparser)
        # ... and tell about how it went
    except DiagnosticError:
        # If there is a diagnostic error, don't say anything, the diagnostics
        # are enough.
        return False
    else:
        print('Code generation was successful')
        return True
    finally:
        lexer._dfa_code = None
        langkit.reset()


def build(grammar, lexer=None, warning_set=default_warning_set):
    """
    Shortcut for `build_and_run` to only build.
    """
    build_and_run(grammar, lexer=lexer, warning_set=warning_set)


def build_and_run(grammar, py_script=None, ada_main=None, lexer=None,
                  ocaml_main=None, warning_set=default_warning_set,
                  generate_unparser=False, symbol_canonicalizer=None):
    """
    Compile and emit code for `ctx` and build the generated library. Then,
    execute the provided scripts/programs, if any.

    An exception is raised if any step fails (the script must return code 0).

    :param langkit.lexer.Lexer lexer: The lexer to use along with the grammar.
        See emit_and_print_errors.
    :param None|str py_script: If not None, name of the Python script to run
        with the built library available.
    :param None|str|list[str] ada_main: If not None, list of name of main
        source files for Ada programs to build and run with the generated
        library. If the input is a single string, consider it's a single mail
        source file.
    :param WarningSet warning_set: Set of warnings to emit.
    :param bool generate_unparser: Whether to generate unparser.
    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    ctx = prepare_context(grammar, lexer, warning_set,
                          symbol_canonicalizer=symbol_canonicalizer)

    class Manage(ManageScript):
        def create_context(self, args):
            return ctx

    m = Manage(override_lang_source_dir=False)

    extensions_dir = os.path.abspath('extensions')
    if os.path.isdir(extensions_dir):
        ctx.extensions_dir = extensions_dir

    # First build the library. Forward all test.py's arguments to the libmanage
    # call so that manual testcase runs can pass "-g", for instance.
    argv = sys.argv[1:] + ['--full-error-traces', '-vnone', 'make']
    for w in WarningSet.available_warnings:
        argv.append('-{}{}'.format('W' if w in warning_set else 'w', w.name))
    if not pretty_print:
        argv.append('--no-pretty-print')
    if generate_unparser:
        argv.append('--generate-unparser')
    m.run(argv)

    # Flush stdout and stderr, so that diagnostics appear deterministically
    # before the script/program output.
    sys.stdout.flush()
    sys.stderr.flush()

    # Write a "setenv" script to make developper investigation convenient
    with open('setenv.sh', 'w') as f:
        m.write_setenv(f)

    env = m.derived_env()

    def run(*argv):
        subprocess.check_call(argv, env=env)

    if py_script is not None:
        # Run the Python script. Note that in order to use the generated
        # library, we have to use the special Python interpreter the testsuite
        # provides us. See the corresponding code in
        # testuite_support/python_driver.py.
        python_interpreter = os.environ['PYTHON_INTERPRETER']
        run(python_interpreter, py_script)

    if ada_main is not None:
        if isinstance(ada_main, str):
            ada_main = [ada_main]

        # Generate a project file to build the given Ada main and then run
        # the program.
        with open('gen.gpr', 'w') as f:
            f.write(project_template.format(
                main_sources=', '.join('"{}"'.format(m) for m in ada_main)
            ))
        run('gprbuild', '-Pgen', '-q', '-p',
            '-XLIBRARY_TYPE=relocatable', '-XXMLADA_BUILD=relocatable')

        for i, m in enumerate(ada_main):
            assert m.endswith('.adb')
            if i > 0:
                print('')
            if len(ada_main) > 1:
                print('== {} =='.format(m))
            sys.stdout.flush()
            argv = [os.path.join('obj', m[:-4])]
            if os.environ.get('VALGRIND_ENABLED'):
                argv = valgrind_cmd(argv)
            run(*argv)

    if ocaml_main is not None:
        run('dune', 'exec', '--display', 'quiet', '--root', '.',
            './{}.exe'.format(ocaml_main))


def add_gpr_path(dirname):
    """
    Prepend the given directory name to the ``GPR_PROJECT_PATH`` environment
    variable.
    """
    old_path = os.environ.get('GPR_PROJECT_PATH')
    os.environ['GPR_PROJECT_PATH'] = (
        '{}{}{}'.format(dirname, os.path.pathsep, old_path)
        if old_path else dirname)
