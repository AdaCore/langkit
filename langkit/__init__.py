def reset():
    """
    Reset global state in Langkit.

    TODO: this is a hack to workaround another hack. At some point in the
    future, we should get rid of this global state in Langkit.
    """

    import langkit.compile_context
    import langkit.compiled_types
    import langkit.expressions.base
    import langkit.lexer
    import langkit.utils

    langkit.utils.reset_memoized()

    langkit.compiled_types.CompiledTypeRepo.reset()
    langkit.compiled_types.create_builtin_types()

    langkit.expressions.base.Entity.unfreeze()
    langkit.expressions.base.Self.unfreeze()

    langkit.lexer.LexerToken.reset()

    langkit.compile_context.compile_ctx = None
