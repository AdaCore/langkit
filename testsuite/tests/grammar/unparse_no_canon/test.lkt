# Test the generation of unparsers in a configuration where there is no single
# cannonical parser that contains all information for unparsing. As a
# consequence, this also checks that unparsers are properly combined.

import lexer_example

@with_lexer(foo_lexer)
@with_unparsers
grammar foo_grammar {
    @main_rule main_rule <- list+(or(
        | RootNode(
            "def"
            null(Identifier) ?pick("{" Number(@Number) "}") ";"
        )
        | RootNode(
            "def"
            ?pick("(" Identifier(@Identifier) ")") null(Number) ";"
        )
    ))
}

@abstract
class FooNode implements Node[FooNode] {
}

class Identifier: FooNode implements TokenNode {
}

class Number: FooNode implements TokenNode {
}

class RootNode: FooNode {
    @parse_field @nullable ident: Identifier
    @parse_field @nullable number: Number
}
