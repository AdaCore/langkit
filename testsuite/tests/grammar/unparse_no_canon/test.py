"""
Test the generation of unparsers in a configuration where there is no single
cannonical parser that contains all information for unparsing. As a
consequence, this also checks that unparsers are properly combined.
"""

from langkit.dsl import ASTNode, Field

from utils import build_and_run


class FooNode(ASTNode):
    pass


class RootNode(FooNode):
    ident = Field()
    number = Field()


class Identifier(FooNode):
    token_node = True


class Number(FooNode):
    token_node = True


build_and_run(lkt_file='expected_concrete_syntax.lkt',
              ada_main='main.adb', generate_unparser=True, types_from_lkt=True)

print('Done')
