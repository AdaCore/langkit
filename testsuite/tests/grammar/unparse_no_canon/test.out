main.adb: starting...
def { 1 } ; def ( a ) ;

main.adb: done.
Done
